File: backend/src/middleware/errorHandler.ts
/**
 * Centralized error handling middleware
 * Provides user-friendly error messages without exposing internal details
 */

import { Request, Response, NextFunction } from "express";
import { ErrorResponse } from "../types/api";

export class AppError extends Error {
  constructor(
    public statusCode: number,
    public message: string,
    public isOperational = true
  ) {
    super(message);
    Object.setPrototypeOf(this, AppError.prototype);
  }
}

export const errorHandler = (
  err: Error | AppError,
  req: Request,
  res: Response,
  next: NextFunction
): void => {
  // Log error details internally
  console.error("Error:", {
    message: err.message,
    stack: err.stack,
    path: req.path,
    method: req.method,
  });

  // Handle known operational errors
  if (err instanceof AppError && err.isOperational) {
    const errorResponse: ErrorResponse = {
      error: getErrorName(err.statusCode),
      message: err.message,
    };
    res.status(err.statusCode).json(errorResponse);
    return;
  }

  // Handle OpenAI API errors
  if (err.message.includes("OpenAI") || err.message.includes("API")) {
    // Map OpenAI errors to appropriate status codes
    if (err.message.includes("401") || err.message.includes("authentication")) {
      res.status(500).json({
        error: "Internal Server Error",
        message: "Authentication failed. Please contact support.",
      });
      return;
    }
    if (err.message.includes("429") || err.message.includes("rate limit")) {
      res.status(503).json({
        error: "Service Unavailable",
        message: "Service is temporarily unavailable. Please try again later.",
      });
      return;
    }
    if (err.message.includes("timeout")) {
      res.status(503).json({
        error: "Service Unavailable",
        message: "Request timed out. Please try again later.",
      });
      return;
    }
    // Generic OpenAI error
    res.status(503).json({
      error: "Service Unavailable",
      message: "Service is temporarily unavailable. Please try again later.",
    });
    return;
  }

  // Handle validation errors (from Zod)
  if (err.name === "ZodError") {
    res.status(400).json({
      error: "Bad Request",
      message: "Invalid request data. Please check your input.",
    });
    return;
  }

  // Default: Internal Server Error
  res.status(500).json({
    error: "Internal Server Error",
    message: "An unexpected error occurred. Please try again later.",
  });
};

function getErrorName(statusCode: number): string {
  switch (statusCode) {
    case 400:
      return "Bad Request";
    case 401:
      return "Unauthorized";
    case 403:
      return "Forbidden";
    case 404:
      return "Not Found";
    case 429:
      return "Too Many Requests";
    case 500:
      return "Internal Server Error";
    case 503:
      return "Service Unavailable";
    default:
      return "Error";
  }
}

File: backend/app/middleware/error_handler.py
"""
Error handling middleware for FastAPI
"""

import logging
from fastapi import Request, status
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError
from openai import APIError
from app.exceptions import AppError
from app.models.responses import ErrorResponse

logger = logging.getLogger(__name__)


async def app_error_handler(request: Request, exc: AppError) -> JSONResponse:
    """
    Handle custom AppError exceptions
    """
    logger.error(f"AppError: {exc.message} (status_code: {exc.status_code})")
    
    return JSONResponse(
        status_code=exc.status_code,
        content=ErrorResponse(
            error=f"HTTP {exc.status_code}",
            message=exc.message
        ).model_dump()
    )


async def validation_error_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    """
    Handle Pydantic validation errors
    """
    errors = exc.errors()
    error_messages = []
    
    for error in errors:
        field = ".".join(str(loc) for loc in error.get("loc", []))
        message = error.get("msg", "Validation error")
        error_messages.append(f"{field}: {message}")
    
    error_message = "; ".join(error_messages) if error_messages else "Validation error"
    logger.warning(f"Validation error: {error_message}")
    
    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content=ErrorResponse(
            error="Validation Error",
            message=error_message
        ).model_dump()
    )


async def openai_error_handler(request: Request, exc: APIError) -> JSONResponse:
    """
    Handle OpenAI API errors
    """
    status_code = 503  # Service Unavailable by default
    
    if exc.status_code == 401:
        logger.error("OpenAI API authentication error")
        status_code = 500  # Internal Server Error (don't expose auth issues)
    elif exc.status_code == 429:
        logger.error("OpenAI API rate limit error")
        status_code = 503
    else:
        logger.error(f"OpenAI API error: {exc.message}")
    
    return JSONResponse(
        status_code=status_code,
        content=ErrorResponse(
            error=f"HTTP {status_code}",
            message="Service temporarily unavailable" if status_code == 503 else "Internal server error"
        ).model_dump()
    )


async def generic_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    """
    Handle unexpected exceptions
    """
    logger.exception(f"Unexpected error: {exc}")
    
    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content=ErrorResponse(
            error="Internal Server Error",
            message="An unexpected error occurred"
        ).model_dump()
    )
